keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,316 @@
+of the state of the node.
+
+As the number of pods on a node increases, the amount of time the Kubelet and CRI implementation takes in generating and reading this list increases linearly. What is needed is a way of the Kubelet being notified when a container or image changes state in a way it did not trigger.
+
+Luckily, there should only be two such cases, and in normal operation, only one would happen frequently:
+
+- The first, and most clear case of a container or image changing state without the Kubelet triggering that state change is when a container stops. Containers can exit gracefully, or be OOM killed, and the Kubelet would not know.
+   - We will also introduce events when the container is created as well as is started. This will help us reduce the relisting that takes placed while the kubelet waits for the container to start.
+   - Although kubelet initiates the container deletion, for sake of increased validation we are also introducing the event to denote that from the runtime.
+- The second, and less likely case is when another entity comes and changes the state of the node.
+	- For container related events (such as a container creating, starting, stopping or being killed), this can appear as a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) manually, or even using the runtime directly.
+	- For image related events (such as an image being pulled, or removed), this can also appear from a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md), or using various image moving tools (podman, nerdctl, skopeo, etc) to move images independent of the Kubelet.
+
+The Kubelet currently covers each of thse cases quite easily: by listing all of the resources on the node, it will have an accurate picture after the amount of time of its [poll interval](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162). For each of these cases, a new CRI-based events API can be made, using [gRPC server streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). This way, the entity closest to the activity of the containers, images and pods (the CRI implementation) can be responsible for informing the Kubelet of their behavior directly.
+
+
+### User Stories
+
+- As a cluster administrator I want to enable `Evented PLEG` feature of the kubelet for better performance.
```

Comment:
I think we should add one where a cluster admin wants their nodes to perform with as little infrastructure overhead as possible